File: mek_stripe_terminal/ios/mek_stripe_terminal/Sources/mek_stripe_terminal/Mappings/TestReaderUpdateMappings.swift
import Foundation
import StripeTerminal

extension TestReaderUpdateApi {
    func toHost() -> TestReaderUpdate {
        let components = (self.components ?? []).toHost()
        switch updateType {
        case .available:
            return TestReaderUpdate.available(components)
        case .required:
            return TestReaderUpdate.required(components)
        case .requiredOffline:
            return TestReaderUpdate.requiredOffline(components)
        case .lowBattery:
            return TestReaderUpdate.lowBattery()
        case .lowBatterySucceedConnect:
            return TestReaderUpdate.lowBatterySucceedConnect()
        }
    }
}

private extension Array where Element == UpdateComponentApi {
    func toHost() -> UpdateComponent {
        var result: UpdateComponent = []
        for component in self {
            switch component {
            case .incremental:
                result.insert(.incremental)
            case .firmware:
                result.insert(.firmware)
            case .config:
                result.insert(.config)
            case .keys:
                result.insert(.keys)
            }
        }
        return result
    }
}

File: mek_stripe_terminal/android/src/main/kotlin/mek/stripeterminal/mappings/TestReaderUpdateMappings.kt
package mek.stripeterminal.mappings

import com.stripe.stripeterminal.external.models.ReaderSoftwareUpdate
import com.stripe.stripeterminal.external.models.TestReaderUpdate
import mek.stripeterminal.api.TestReaderUpdateApi
import mek.stripeterminal.api.TestReaderUpdateTypeApi
import mek.stripeterminal.api.UpdateComponentApi

fun TestReaderUpdateApi.toHost(): TestReaderUpdate {
    val components = (components ?: emptyList()).map { it.toHost() }.toSet()
    return when (updateType) {
        TestReaderUpdateTypeApi.AVAILABLE -> TestReaderUpdate.available(components)
        TestReaderUpdateTypeApi.REQUIRED -> TestReaderUpdate.required(components)
        TestReaderUpdateTypeApi.REQUIRED_OFFLINE -> TestReaderUpdate.requiredOffline(components)
        TestReaderUpdateTypeApi.LOW_BATTERY -> TestReaderUpdate.lowBattery()
        TestReaderUpdateTypeApi.LOW_BATTERY_SUCCEED_CONNECT -> TestReaderUpdate.lowBatterySucceedConnect()
    }
}

fun UpdateComponentApi.toHost(): ReaderSoftwareUpdate.UpdateComponent {
    return when (this) {
        UpdateComponentApi.INCREMENTAL -> ReaderSoftwareUpdate.UpdateComponent.INCREMENTAL
        UpdateComponentApi.FIRMWARE -> ReaderSoftwareUpdate.UpdateComponent.FIRMWARE
        UpdateComponentApi.CONFIG -> ReaderSoftwareUpdate.UpdateComponent.CONFIG
        UpdateComponentApi.KEYS -> ReaderSoftwareUpdate.UpdateComponent.KEYS
    }
}
